src/distPlus/symbolic/SymbolicDist.re

```diff
@@ -296,103 +313,407 @@ module GenericSimple = {
+    }
+  }
+
+  let renderNode = (et: nodeResult, n: int) => {
+    switch (et) {
+    | `Simple(d) => {
+          let (sc, sd) = renderDistributionToXYShape(d, n);
+          `RenderedShape(sc, sd, 1.0);
+      }
+    | s => s
+    }
+  }
+
+  let rec evaluateNode = (treeNode: distTree, n: int): nodeResult => {
+    // returns either a new symbolic distribution
+    switch (treeNode) {
+    | `Simple(d) => evaluateDistribution(d)
+    | `Combination(t1, t2, op) => evaluateCombinationDistribution(evaluateNode(t1, n), evaluateNode(t2, n), op, n)
+    | `PointwiseSum(t1, t2) => evaluatePointwiseSum(evaluateNode(t1, n), evaluateNode(t2, n), n)
+    | `PointwiseProduct(t1, t2) => evaluatePointwiseProduct(evaluateNode(t1, n), evaluateNode(t2, n), n)
+    | `VerticalScaling(t1, t2) => evaluateVerticalScaling(evaluateNode(t1, n), evaluateNode(t2, n), n)
+    | `Normalize(t) => evaluateNormalize(evaluateNode(t, n), n)
+    | `LeftTruncate(t, x) => evaluateTruncate(evaluateNode(t, n), x, (>=), n)
+    | `RightTruncate(t, x) => evaluateTruncate(evaluateNode(t, n), x, (<=), n)
+    | `Render(t) => renderNode(evaluateNode(t, n), n)
+    }
+  };
+
+  let toShape = (treeNode: distTree, n: int) => {
+    let treeShape = evaluateNode(`Render(`Normalize(treeNode)), n);
+
+    switch (treeShape) {
+    | `Simple(_) => E.O.toExn("No shape found!", None)
+    | `RenderedShape(sc, sd, _) => {
+      let shape = MixedShapeBuilder.buildSimple(~continuous=Some(sc), ~discrete=sd);
+
+      shape |> E.O.toExt("");
+    }
+    }
+  };
+
+  let rec toString = (treeNode: distTree): string => {
+    let stringFromOp = op => switch (op) {
+      | `AddOperation => " + "
+      | `SubtractOperation => " - "
+      | `MultiplyOperation => " * "
+      | `DivideOperation => " / "
+      | `ExponentiateOperation => "^"
+    };
+
+    switch (treeNode) {
     | `Simple(d) => GenericSimple.toString(d)
-    | `PointwiseCombination(d) =>
-      PointwiseAddDistributionsWeighted.toString(d)
-  );
-
-let toShape = n =>
-  fun
-  | `Simple(d) => GenericSimple.toShape(~xSelection=`ByWeight, d, n)
-  | `PointwiseCombination(d) =>
-    PointwiseAddDistributionsWeighted.toShape(d, n);
+    | `Combination(t1, t2, op) => toString(t1) ++ stringFromOp(op) ++ toString(t2)
+    | `PointwiseSum(t1, t2) => toString(t1) ++ " .+ " ++ toString(t2)
+    | `PointwiseProduct(t1, t2) => toString(t1) ++ " .* " ++ toString(t2)
+    | `VerticalScaling(t1, t2) => toString(t1) ++ " @ " ++ toString(t2)
+    | `Normalize(t) => "normalize(" ++ toString(t) ++ ")"
+    | `LeftTruncate(t, x) => "leftTruncate(" ++ toString(t) ++ ", " ++ string_of_float(x) ++ ")"
+    | `RightTruncate(t, x) => "rightTruncate(" ++ toString(t) ++ ", " ++ string_of_float(x) ++ ")"
+    | `Render(t) => toString(t)
+    }
+  };
+};
+
+let toString = (treeNode: distTree) => DistTree.toString(treeNode)
+
+let toShape = (sampleCount: int, treeNode: distTree) =>
```

Comment:
what do you mean by ``treeNode``? Can we just call this ``distTree`` maybe?

Reply:
Ah, I didn't realize you were allowed to use type names like that. Done